src/interface/widgets/STPVisualizerWidget.cpp

```diff
@@ -0,0 +1,142 @@
+/*
+ *
+ * This widget visualizes a behaviour (strategy) tree.
+ *
+ * Update contents updates the statuses, ticktimes and amount of ticks for each node.
+ * The nodes in the tree are not refreshed until the whole tree is invalidated.
+ *
+ */
+
+#include "stp/Play.hpp"
+#include "stp/Role.hpp"
+#include "interface/widgets/STPVisualizerWidget.h"
+#include <treeinterp/BTFactory.h>
+#include <QtWidgets/QLayoutItem>
+#include <include/roboteam_ai/bt/Node.h>
+#include "interface/widgets/mainWindow.h"
+
+namespace rtt::ai::interface {
+    constexpr static const char *space = "&nbsp;";
+    constexpr static const char *tab = "&nbsp;&nbsp;&nbsp;&nbsp;";
+
+    STPVisualizerWidget::STPVisualizerWidget(MainWindow *parent) : QTextEdit(parent) {
+        this->parent = parent;
+        this->setReadOnly(true);
+    }
+
+    std::string_view STPVisualizerWidget::getColorForStatus(stp::Status status) {
+        switch (status) {
+            case stp::Status::Failure:
+                return "<font color=\"Red\">";
+            case stp::Status::Running:
+                return "<font color=\"Green\">";  // dark green
+            case stp::Status::Success:
+                return "<font color=\"Lime\">";  // bright green
+            case stp::Status::Waiting:
+                return "<font color=\"Aqua\">";
+            default:
+                return "<font color=\"White\">";
+        }
+    }
+
+    void STPVisualizerWidget::updateContents(stp::Play *currentPlay) {
+        std::lock_guard lck{ contentLock };
+        updateContent.str("");
+        displayPlay(currentPlay);
+    }
+
+    void STPVisualizerWidget::displayPlay(stp::Play *currentPlay) {
+        updateContent << "Play: ";
+        if (!currentPlay) {
+            updateContent << "None<br>";
+            return;
+        }
+        updateContent << "Some(" << currentPlay->getName() << ")<br>" << tab;
+        std::vector<std::pair<stp::Role*, stp::Status>> states = { currentPlay->getRoleStatuses().begin(), currentPlay->getRoleStatuses().end() };
+        std::sort(states.begin(), states.end(), [](auto const& lhs, auto const& rhs) {
```

Comment:
Maybe more descriptive names for the lhs and rhs of the hashmap, but that's minor